frontend/ui/widget/spinwidget.lua

```diff
@@ -167,6 +147,29 @@ function SpinWidget:update()
             },
         })
     end
+    table.insert(buttons, {
+        {
+            text = self.cancel_text,
+            callback = function()
+                if self.cancel_callback then
+                    self.cancel_callback()
+                end
+                self:onClose()
+            end,
+        },
+        {
+            text = self.ok_text,
+            callback = function()
+                if self.callback then
+                    self.value, self.value_index = value_widget:getValue()
+                    self.callback(self)
```

Comment:
Maybe just a nit. Why don't use `self:callback()`?

Reply:
I was going to agree (indeed, `self:callback()` is the classic shorten form of `self.callback(self)` ).
But I think our "callback" values are just plain functions/callables, so you just call them, and there, it happens to be stored in `self`.
And the fact that you pass `self` as an argument is ... a coincidence ? :) It could usually be any value, or a `touchmenu_instance`, or anything - but here, it's `self`.
So, it might be more explicite written as `self.callback(self)`.
(Haven't checked the context, just some thoughts while reading these 4 lines.)

Reply:
Have not checked other widgets either, but as a shoot in the dark: Would it be a problem to call **all** other callbacks with self (aka `:`) and touchmenu_instance (`self:xxxcallback(touchmenu_instance)`). This would have saved me quite some time on the datemenuwidget/autosuspend PR

Reply:
I guess it would be a problem: these callbacks are not methods (functions of a class that take an instance of that class as their first argument, which with the `:` syntactic sugar happens to be known as `self` inside the code), but just callable (we define them as simple functions, in a context that may have nothing to do with a class or instance - or may, an in there, `self` is some upper value that we can use or not).
The time you lost is probably because you did not know (yet) the difference (happened to me too) - but now you won't any longer :) (I admit it's quite a short explanation, but you'd have to look at some code to understand yourself the difference.)

Reply:
> (we define them as simple functions, in a context that may have nothing to do with a class or instance

Agreed, that is the current behaviour. But the question arises: Would it do any harm if we change our current behaviour and give the poor homeless functions a hook to link in, if they want? 
Current functions don't get any parameters so they wont mind them, but future functions can take a profit of the implicite self and the touchmenu_instance. This sounds  like  a no loose but win situation.
What do you think?

Reply:
@poire-z The time I lost, was a time to learn :))))) I have advanced. ++

That is part of the business. My concern is about the next generation of contributors, the next PRs.Can we make things smoother?

Reply:
Dunno. It seems to me we do things the classic Lua way (+ our :new / :init stuff, that may not be standard and is to be discovered and understood).
But any contributor should have a look at the Lua docs and the Lua way of doing functions and closures, and class and objects:
https://www.lua.org/pil/6.html
https://www.lua.org/pil/6.1.html
https://www.lua.org/pil/16.html

> Current functions don't get any parameters so they wont mind them, but future functions can take a profit of the implicite self and the touchmenu_instance. This sounds like a no loose but win situation.
What do you think?

I think not :) The self you anchor/access the function in/from may not be the self you expect from inside the function. Just try to do it to some of the other callbacks, and you may find it a messy-messy situation. (Most of the time, you don't have any `self` around - and if you define the callback inside a method, you have the `self` object anyway available as an upvalue (the callback you define is a function, and the values defines outside of it are accessable, and will live even if the `self` is gone and nothing references it anymore - the callback still references it - so you may not even need to bother with storing and passing a `self`. There's some kind of natural beauty in all this ;) that you want to add make up too :) But it doesn't' need it, one just has to play and stumble on it to discover how it works.

(An other thing that bugged me initially, is that in Python, a bound method is an object you can pass around - and Lua doesn't have that - but with the closure/upvalue stuff, it doesn't matter.)